lib/common/ipc_schedulerd.c

```diff
@@ -0,0 +1,178 @@
+{
+    free(((struct schedulerd_api_private_s *) data)->client_uuid);
+    free(data);
+}
+
+// \return Standard Pacemaker return code
+static int
+post_connect(pcmk_ipc_api_t *api)
+{
+    if (api->api_data == NULL) {
+        return EINVAL;
+    }
+
+    return pcmk_rc_ok;
+}
+
+static bool
+reply_expected(pcmk_ipc_api_t *api, xmlNode *request)
+{
+    const char *command = crm_element_value(request, F_CRM_TASK);
+
+    if (command == NULL) {
+        return false;
+    }
+
+    // We only need to handle commands that functions in this file can send
+    return pcmk__str_any_of(command, CRM_OP_PECALC, NULL);
+}
+
+static void
+dispatch(pcmk_ipc_api_t *api, xmlNode *reply)
+{
+    crm_exit_t status = CRM_EX_OK;
+    xmlNode *msg_data = NULL;
+    pcmk_schedulerd_api_reply_t reply_data = {
+        pcmk_schedulerd_reply_unknown
+    };
+    const char *value = NULL;
+    long long value_ll = 0;
+
+    if (pcmk__str_eq((const char *) reply->name, "ack", pcmk__str_casei)) {
+        return;
+    }
+
+    value = crm_element_value(reply, F_CRM_MSG_TYPE);
+    if ((value == NULL) || (strcmp(value, XML_ATTR_RESPONSE))) {
+        crm_debug("Unrecognizable schedulerd message: invalid message type '%s'",
+                  crm_str(value));
+        status = CRM_EX_PROTOCOL;
+        goto done;
+    }
+
+    if (crm_element_value(reply, XML_ATTR_REFERENCE) == NULL) {
+        crm_debug("Unrecognizable schedulerd message: no reference");
+        status = CRM_EX_PROTOCOL;
+        goto done;
+    }
+
+    value = crm_element_value(reply, F_CRM_TASK);
+
+    // Parse useful info from reply
+    msg_data = get_message_xml(reply, F_CRM_DATA);
+    crm_element_value_ll(msg_data, XML_ATTR_TSTAMP, &value_ll);
+
+    if (pcmk__str_eq(value, CRM_OP_PECALC, pcmk__str_none)) {
+        reply_data.reply_type = pcmk_schedulerd_reply_graph;
+        reply_data.data.graph.reference = crm_element_value(reply, XML_ATTR_REFERENCE);
+        reply_data.data.graph.input = crm_element_value(reply, F_CRM_TGRAPH_INPUT);
+        reply_data.data.graph.tgraph = msg_data;
+    } else {
+        crm_debug("Unrecognizable pacemakerd message: '%s'", crm_str(value));
+        status = CRM_EX_PROTOCOL;
+        goto done;
+    }
+
+done:
+    pcmk__call_ipc_callback(api, pcmk_ipc_event_reply, status, &reply_data);
+}
+
+pcmk__ipc_methods_t *
+pcmk__schedulerd_api_methods()
+{
+    pcmk__ipc_methods_t *cmds = calloc(1, sizeof(pcmk__ipc_methods_t));
+
+    if (cmds != NULL) {
+        cmds->new_data = new_data;
+        cmds->free_data = free_data;
+        cmds->post_connect = post_connect;
+        cmds->reply_expected = reply_expected;
+        cmds->dispatch = dispatch;
+    }
+    return cmds;
+}
+
+static int
+do_schedulerd_api_call(pcmk_ipc_api_t *api, const char *ipc_name, const char *task,
+                       xmlNode *xml_data, char **ref)
+{
+    schedulerd_api_private_t *private;
+    xmlNode *cmd = NULL;
+    int rc;
+
+    CRM_CHECK(api != NULL, return -EINVAL);
```

Comment:
Since this (actually the caller) is public API, I think a regular if is better than CRM_CHECK()

Also + not -